keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -2236,6 +2197,10 @@ message ContainerConfig {
 }
 ```
 
+At the time of Kubernetes 1.31, most container runtimes [support this
+field](https://github.com/kubernetes/kubernetes/issues/125210) so it can be
+used instead of some earlier approach with annotations.
+
```

Comment:
Is `NodePrepareResource` called at an appropriate time to, for example, add a chosen NIC into the network namespace of the pod sandbox? NICs are not added to container specs, I want to make sure that on the kubelet side, we have an appropriate call to allow pod-scoped devices to get activated. "Appropriate" means both at the right time, *and* with the right inputs (the allocations, in particular). Based on what I am reading here, the sandbox is not ready yet, and this would not have access to the network namespace name. We may need another call?

cc @aojea @klueska

Reply:
It is called here as part of sandbox creation itself:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1140

We added it here on specific request by the networking team at NVIDIA (though I was under the impression that this was done just to ensure it happened strictly *before* individual container creation -- meaning we could have also put it in the pod admission loop if that were an option).

It sounds like you actually want it later in the sandbox creation step though. I don't see any obvious problems with that -- I would just need to understand where exactly and what why.

Reply:
/cc @moshe010 since he has the most context as to why we put it here where we did

Reply:
We need better documentation of the actual behavior and the rationale behind it in the KEP. We treated it as an implementation detail before, but it clearly has design implications that should be surfaced here.

This is not a blocker for this PR, just a general comment.

Reply:
This is perfect , it has to happen *before* sandbox creation , the Network DRA driver has to have another leg in the runtime, the implementation there can be CNI, CDI (OCI hooks) or NRI (more details in https://docs.google.com/presentation/d/1pjDCtpdbCSWaqCbBYWgzTxAewOVbMf6rUS5SbjAJAe8/edit#slide=id.g2c2d098b0a2_0_481)

This hook before creating the sandbox will allow the driver to prepare whatever is necessary so the callback on the runtime will have the necessary information

Reply:
Still worth documenting this in the KEP. @pohly can you take care of that?

Reply:
I think you are better suited to do that. Doesn't need to be in this PR.

Reply:
Yes, so we wanted the PrepareResources called before invoking the CNI (before it was on the container level) 
Here is the issue [1] and fix [2] 

[1] - https://github.com/kubernetes/kubernetes/issues/113785
[2] - https://github.com/kubernetes/kubernetes/pull/114364

Reply:
Related to this: should or should not NodePrepareResources be called for claims which are not referenced by any container?

With networking in mind, the answer is probably "yes", but I vaguely remember that the current code skips the call.

Reply:
You remember correctly, the current code skips preparing claims that are not referenced by any container. This will either need to change, or a `pause` container will need to be used to consume such resources.

Reply:
I think that should change. The user associated the claim with the pod; let's let the driver decide if anything needs to be done with it.

Reply:
I tend to agree. As it's not documented here, I'm going to treat it as an implementation detail that we can change for 1.31.